File: BEmuJava/bemu/src/main/java/com/bloomberglp/blpapi/MarketElementExceptions.java
//------------------------------------------------------------------------------
// <copyright project="BEmu_maven" file="/BEmu_maven/bemu/src/main/java/com/bloomberglp/blpapi/MarketElementExceptions.java" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

package com.bloomberglp.blpapi;

import com.bloomberglp.blpapi.Element;
import com.bloomberglp.blpapi.Name;
import com.bloomberglp.blpapi.SchemaTypeDefinition;

public class MarketElementExceptions extends Element
{
    private final MarketElementString _fieldId;
    private final MarketElementReason _reason;

    public MarketElementExceptions(String badField)
    {
        this._fieldId = new MarketElementString("fieldId", badField);
        this._reason = new MarketElementReason(MarketElementReason.ReasonTypeEnum.badField);
    }
    
    public SchemaTypeDefinition typeDefinition() throws Exception
    {
    	return new SchemaTypeDefinition(this.datatype(), new Name("SubscriptionException"));
    }
	
    public Name name() throws Exception
    {
    	return new Name("exceptions");
    }
    
    public int numValues()
    {
    	return 1;
    }
    
    public int numElements()
    {
    	return 2;
    }
    
    public boolean isComplexType()
    {
    	return true;
    }
    
    public boolean isArray()
    {
    	return false;
    }
    
    public boolean isNull()
    {
    	return false;
    }
    
    public boolean hasElement(String name, boolean excludeNullElements)
    {
    	return this.hasElement(name);
    }
    
    public boolean hasElement(String name)
    {
    	return name.equals("fieldId") || name.equals("reason");
    }

    public String getElementAsString(String name) throws Exception
    {
        return this.getElement(name).getValueAsString();
    }
    
    public Element getElement(String name) throws Exception
    {
    	if(name.equals("fieldId"))
    		return this._fieldId;
    	
    	else if(name.equals("reason"))
    		return this._reason;
    	
    	else
    		throw new Exception("name not recognized. names are case-sensitive.");
    }
    
    protected StringBuilder prettyPrint(int tabIndent) throws Exception
    {
        String tabs = com.bloomberglp.blpapi.IndentType.Indent(tabIndent);
        StringBuilder result = new StringBuilder();
        
        result.append(String.format("%s%s = {%s", tabs, this.name().toString(), System.getProperty("line.separator")));
        result.append(this._fieldId.prettyPrint(tabIndent + 1));
        result.append(this._reason.prettyPrint(tabIndent + 1));
        result.append(String.format("%s}%s", tabs, System.getProperty("line.separator")));

        return result;
    }

}

File: BEmu_cpp/src/MarketDataRequest/MarketElementExceptions.cpp
//------------------------------------------------------------------------------
// <copyright project="BEmu_cpp" file="src/MarketDataRequest/MarketElementExceptions.cpp" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

#include "MarketDataRequest/MarketElementExceptions.h"
#include "MarketDataRequest/MarketElementString.h"
#include "MarketDataRequest/MarketElementReason.h"

#include "Types/IndentType.h"

namespace BEmu
{
	namespace MarketDataRequest
	{
		MarketElementExceptions::MarketElementExceptions(const std::string& badField) :
			_fieldId(new MarketElementString("fieldId", badField)),
			_reason(new MarketElementReason(MarketDataRequest::ReasonType::badField))
		{
		}

		MarketElementExceptions::~MarketElementExceptions()
		{
		}

		Name MarketElementExceptions::name() const
		{
			return Name("exceptions");
		}

		size_t MarketElementExceptions::numValues() const
		{
			return 1;
		}

		size_t MarketElementExceptions::numElements() const
		{
			return 2;
		}

		SchemaElementDefinition MarketElementExceptions::elementDefinition() const
		{
			::blpapi_DataType_t dtype = (::blpapi_DataType_t)this->datatype();
			SchemaElementDefinition result(dtype, Name("SubscriptionException"));
			return result;
		}

		bool MarketElementExceptions::isNull() const
		{
			return false;
		}

		bool MarketElementExceptions::isArray() const
		{
			return false;
		}

		bool MarketElementExceptions::isComplexType() const
		{
			return true;
		}

		boost::shared_ptr<ElementPtr> MarketElementExceptions::getElement(const char* name) const
		{
			if(strncmp(name, "fieldId", 8) == 0)
				return boost::dynamic_pointer_cast<ElementPtr>(this->_fieldId);
			
			else if(strncmp(name, "reason", 7) == 0)
				return boost::dynamic_pointer_cast<ElementPtr>(this->_reason);

			else
				throw elementPtrEx;
		}

		bool MarketElementExceptions::hasElement(const char* name, bool excludeNullElements) const
		{
			return strncmp(name, "fieldId", 8) == 0 || strncmp(name, "reason", 7) == 0;
		}

		const char* MarketElementExceptions::getElementAsString(const char* name) const
		{
			return this->getElement(name)->getValueAsString(0);
		}

		std::ostream& MarketElementExceptions::print(std::ostream& stream, int level, int spacesPerLevel) const
		{
			std::string tabs(IndentType::Indent(level, spacesPerLevel));

			stream << tabs << "exceptions = {" << std::endl;
			
			this->_fieldId->print(stream, level + 1);
			this->_reason->print(stream, level + 1);

			stream << tabs << '}' << std::endl;

			return stream;
		}

	}
}